### PR TITLE
Add hand-powered battery charger

### PR DESCRIPTION
**Pull request by esotericist**
#### Summary
`SUMMARY: Content "Add hand-powered battery charger"`

#### Purpose of change
With the new-found proliferation of rechargeable batteries, players are somewhat stymied by the lack of options for putting more charge into the batteries. Currently the only available option in-game is the vehicle-mounted Recharging Station.

Partially addresses #30256  (I still think we need additional options, so that issue should probably not be closed yet)

#### Describe the solution
Added a hand-crank charger item, an `iuse` activity, and all the attendant c++ and json plumbing to make it so.

My research suggested that hand-crank charger outputs range from 2 watt (very common and easily verified) to 10 watt (claims I could only find from specific manufacturers, with little outside corroboration). I ultimately opted to aim for just under 7 watts of throughput to get 25 kJ / hour.

At this rate one battery charge is produced every 2.4 minutes (24 turns under the current 6-second turn model), and with the fatigue rate available to me, it looks like a 100% fully rested (zero fatigue) player hits 'dead tired' after about 10.5 hours of constant cranking, yielding ~259 charge. This means a player can come reasonably close to filling a medium battery cell in a single day's labor. The right comestibles could extend this.

Is this worthwhile? Probably not outside of emergencies. But it's an option that was missing that we reasonably should have had.

I'm not 100% sure on the recipe requirements. I was aiming for "substantially easier than the vehicle charger, but not trivial". Feedback welcome.

For worldgen, I added this to `supplies_electronics`, `tools_common`, `tools_electronics`, and `tools_survival`. I'm not very confident on the weights I used there, either.

Basically, if there's any changes someone wants to see on anything except the charge rate (which I'm fairly certain of), I'd be happy to accommodate. 

#### Describe alternatives you've considered
I had originally planning on making a charger that would funnel power from battery magazine A into battery magazine B, but I decided this would be easier to do. I'd still like to get that done at some point.

I considered adding a morale penalty or kcal consumption to this activity, but I decided I just don't have the confidence to work out what those should be.

#### Additional context

image

image

image

### CODE REVIEW

File: data/json/item_actions.json [hunk 1]
```diff
@@ -879,6 +879,11 @@
     "id": "BLOOD_DRAW",
     "name": "Draw some blood"
   },
+  {
+    "type": "item_action",
+    "id": "HAND_CRANK",
+    "name": "Recharging a battery"
```
Comment (Night-Pryanik) [inline data/json/item_actions.json]:
```suggestion
    "name": "Recharge a battery"
```

File: data/json/items/tools.json [hunk 2]
```diff
@@ -8307,6 +8307,35 @@
     "use_action": "BLOOD_DRAW",
     "flags": [ "SPEAR" ]
   },
+  {
+    "id": "hand_crank_charger",
+    "type": "TOOL",
+    "category": "tools",
+    "name": "hand-crank charger",
+    "description": "This is a hand-powered battery charger. It has an adjustable receptacle designed to accept a wide variety of rechargable battery cells.",
```
Comment (Night-Pryanik) [inline data/json/items/tools.json]:
```suggestion
    "description": "This is a hand-powered battery charger.  It has an adjustable receptacle designed to accept a wide variety of rechargeable battery cells.",
```

File: src/activity_handlers.cpp [hunk 3]
```diff
@@ -2023,6 +2024,30 @@ void activity_handlers::vehicle_finish( player_activity *act, player *p )
...
+    // Hand-crank chargers seem to range from 2 watt (very common easily verified)
+    // to 10 watt (suspicious claims from some manufacturers) sustained output.
+    // It takes 2.4 minutes to produce 1kj at just slightly under 7 watts (25 kj per hour)
+    // time-based instead of speed based because it's a sustained activity
+    act->moves_left -= 100;
+    item &hand_crank_item = p ->i_at( act->position );
+
+    // TODO: This should be 144 seconds, rather than 24 (6-second) turns
+    // but we don't have a seconds time macro?
+    if( calendar::once_every( 24_turns ) ) {
```
Comment (Night-Pryanik) [inline src/activity_handlers.cpp]: You can change this to 
```suggestion
    if( calendar::once_every( 144_seconds ) ) {
```

File: src/iuse.cpp [hunk 4]
```diff
@@ -4210,6 +4210,41 @@ int iuse::portable_game( player *p, item *it, bool, const tripoint & )
...
+                                                 it->tname(), it->magazine_current()->tname() ) ) ;
+            p->assign_activity( activity_id( "ACT_HAND_CRANK" ), time, -1, p->get_item_position( it ),
+                                "hand-cranking" );
+        } else {
+            p->add_msg_if_player( string_format(
+                                      _( "You could use the %s to charge its %s, but it's already charged." ), it->tname(),
+                                      magazine->tname() ) ) ;
+        }
+    } else {
+        p->add_msg_if_player( m_info, _( "You need a rechargable battery cell to charge." ) );
```
Comment (Night-Pryanik) [inline src/iuse.cpp]:
```suggestion
        p->add_msg_if_player( m_info, _( "You need a rechargeable battery cell to charge." ) );
```

File: data/json/items/tools.json (hunk 2 above)
Comment (esotericist) [reply data/json/items/tools.json]: I have SO much trouble with that word, my brain insists there must only be one `e` in it. Thanks for spotting that.

File: src/activity_handlers.cpp (hunk 3 above)
Comment (esotericist) [reply src/activity_handlers.cpp]: I swear I tried that and got an error. Let me try again.

Comment (esotericist) [reply src/activity_handlers.cpp]:
```
src/activity_handlers.cpp:2038:34: error: no matching literal operator for call to 'operator""_seconds' with argument of type 'unsigned long long' or 'const char *', and no matching literal operator template
    if( calendar::once_every( 144_seconds ) ) {
                                 ^
```

Both VS Code and clang say this is not legal.

Comment (esotericist) [reply src/activity_handlers.cpp]: That problem is the origin of the comment immediately above the line, btw.

File: data/json/items/tools.json (hunk 2 above)
Comment (Night-Pryanik) [reply data/json/items/tools.json]: You're welcome. That's mostly not me, but spellchecker in my browser.

File: src/activity_handlers.cpp (hunk 3 above)
Comment (Night-Pryanik) [reply src/activity_handlers.cpp]: Well, we could circumvent that by using something like `2_minutes`.

**Comment by adamkad1** [comment]
I like

**Comment by JeanLucVanDamme** [comment]
Seems like the right balance of useful and annoying-to-use. I'm for it.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/how-to-recharge-the-new-battery-cells/20081/2
